### PR TITLE
[9.x] Skip guarding model attributes if the data passed is ValidatedData

### PR DESCRIPTION
**Pull request by themsaid**
This PR skips mass assignment protection if the data passed is an instance of `ValidatedData`.

It covers the following methods:

```php
Model::make();
Model::firstOrNew();
Model::firstOrCreate();
Model::updateOrCreate();
Model::create();
Model::forceCreate();
Model::fill();
Model::forceFill();
Model::update();
Model::updateQuietly();
```

Update:

Relationship methods (create, firstOrCreate, firstOrNew, updateOrCreate) will also accept `ValidatedData` data.

### CODE REVIEW

**Comment by themsaid** [comment]
@taylorotwell Relationship methods (create, firstOrCreate, firstOrNew, updateOrCreate) will also accept ValidatedData data.

**Comment by garygreen** [comment]
I like this idea, but not it's implementation.

It's basically guarded attributes re-imagined, which is a nice problem to solve because fillable/guarded attributes at the moment is too generic to be useful. It really depends on context of the request - what should be allowed to be updated; who's doing it, what's doing it, etc.

So this `ValidatedData` approach of whitelisting attributes on a per-context basis makes sense and should be preferred over the fillable/guarded fields.

However, this current implementation feels clunky having to type-check in every method. Also it invites questions on why all query builder method don't support validated data?

```php
->where(ValidatedData $data)
->orWhere(ValidatedData $data)
->firstOrFail(ValidatedData $data)
```

...etc

So I think this needs to have a clear design goal: only support on "creating/updating"-like methods. And if it can be implemented in a cleaner way without the type checks in every method that would be great.

Also the `ValidatedData` class really needs to nail down security aspect.

**Comment by mcadet** [comment]
What about a class like so? Which will useful in form requests classes, or elsewhere.

```php
<?php

namespace App\Http\Validation;


class PostRequest extends ModelValidation // to implement
{
    public function commonRules() : array
    {
        return [
            'title' => 'required'
        ];
    }

    public function updateRules(): array
    {
        // . . .
    }

    public function storeRules() : array
    {
        return [
            'status' => 'required|in:draft,review'
        ];
    }

    // . . .
}
```

**Comment by taylorotwell** [comment]
@bjhijmans 

```php
$request->safe()->except(['users']);
```

**Comment by garygreen** [comment]
@taylorotwell Your suggestion seems fragile and potentially error prone, more so than fillable fields. I can easily see a scenario there where you validate a field but accidentally forget to "blacklist" it in the `except` list when creating a model.

That's the problem with supporting `ValidatedData` in models - it's so tied to the request it contains all kinds of data not relevant to the model being saved. This means you would have to `except` every attribute not relevant to the model. This could actually expose to *more* security and scalability issues than fillable fields do at the moment.

**Comment by Jubeki** [comment]
@garygreen I think it has many advantages. Some lazy people just deactivate guarding attributes alltogether and all others need to update their fillable property. If you forget to whitelist a field in the fillable property you have also a big problem, because you get a SQL error that the field cannot be null even though it is validated.

In my opinion Transforming the validated data for the needs of the Model should happen in the controller or an action.

**Comment by garygreen** [comment]
@Jubeki

> Some lazy people just deactivate guarding attributes alltogether

It's good practise imo to set fillable fields. I wouldn't advocate something to cater for laziness. Setting fillable fields adds value in knowing the shape of your objects too. Laravel doesn't support [typed properties](https://stitcher.io/blog/typed-properties-in-php-74) models (which would be awesome!) so best we got right now to know what an object looks like is fillable attributes.

> If you forget to whitelist a field in the fillable property you have also a big problem, because you get a SQL error that the field cannot be null even though it is validated.

Once fillable attributes are set, they stay the same. It's the shape of your object and database schema. It doesn't change. The thing that does change though is what values can be saved for each request, who by, where, under what circumstances - context is key.

> In my opinion Transforming the validated data for the needs of the Model should happen in the controller or an action.

This PR essentially allows the active record layer to mingle with your controller + request layer by proxy `ValidatedData`. That comes with its own problems as mentioned above in [this comment](https://github.com/laravel/framework/pull/38427#issuecomment-919452294)

**Comment by Jubeki** [comment]
@garygreen 

> It's good practise imo to set fillable fields. I wouldn't advocate something to cater for laziness. Setting fillable fields adds value in knowing the shape of your objects too. Laravel doesn't support [typed properties](https://stitcher.io/blog/typed-properties-in-php-74) models (which would be awesome!) so best we got right now to know what a model looks like is fillable attributes.

While it may be good practices to maintain a fillable properties, I like to use the docblock above the class for that and generate the docblock using ide-helper with columns from the database.

@themsaid posted a video last year explaining exactly what I am thinking: https://www.youtube.com/watch?v=XNy0hldKHjM

> Once fillable attributes are set, they stay the same until schema changes. It's the shape of your object and database schema. It doesn't change. The thing that does change though is what values can be saved for each request, who by, where, under what circumstances - context is key.

One more reason why I think the transformation and protection from user input should happen in the controller layer and not in the model/database layer.

**Comment by garygreen** [comment]
> While it may be good practices to maintain a fillable properties, I like to use the docblock above the class for that and generate the docblock using ide-helper

The docblocks IDE helper is good but only for that reason... IDE. It's useful to to the developer, but not useful for any runtime features, security and protecting the shape of your objects. 

I would like to see Laravel add better support for typed properties. [I tweeted about this](https://twitter.com/garreh/status/1146457915273990144) a while ago.

image

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -428,51 +429,79 @@ public function findOrNew($id, $columns = ['*'])
...
+     *
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $attributes
+     * @param  array|\Illuminate\Contracts\Support\ValidatedData  $values
+     * @return bool
+     */
+    protected function shouldForceAttributes($attributes, $values)
+    {
+        return $attributes instanceof ValidatedData && $values instanceof ValidatedData ||
+            ($attributes instanceof ValidatedData && empty($values)) ||
+            ($values instanceof ValidatedData && empty($attributes));
```
Comment (inxilpro) [inline src/Illuminate/Database/Eloquent/Builder.php:502]: We could address some of the concerns brought up by only forcing validated data if `fillable` and `guarded` have not been set for this model. Something like:

```suggestion
        return $this->fillable === []
            && $this->guarded === ['*']
            && ($attributes instanceof ValidatedData || empty($attributes))
            && ($values instanceof ValidatedData || empty($values);
```

**Comment by garygreen** [comment]
> We could address some of the concerns brought up by only forcing validated data if fillable and guarded have not been set for this model.

That means people who use fillable and guarded fields won’t be able to use `ValidatedData`. It also doesn't solve the problem of the data containing things not relevant to the model being saved - you'd either have to use `only` or `except` or hope that it only contains data relevant to the model. Either way you've kind of lost the whole design purpose of this feature 🤔

**Comment by mpyw** [comment]
Do not abuse easiness. The feature and behavior should be kept simple and clear.

**Comment by taylorotwell** [comment]
Since this has such mixed reviews going to table this for now.

**Pull request closed without merging**